### PR TITLE
Move iterator implementations to `utils` pkg

### PR DESCRIPTION
**Pull request by dhrubabasu**
## Why this should be merged

This doesn't need to live in the `state` pkg. Let's move it to `utils`. Will be re-used for #3287 

## How this works

Move pkg + update refs

## How this was tested

existing UTs

### CODE REVIEW

File: utils/iterator/empty.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+package iterator
+
+import "github.com/ava-labs/avalanchego/utils"
+
+// Empty returns an iterator with no elements.
+func Empty[T any]() Iterator[T] {
+	return &empty[T]{}
+}
+
+type empty[T any] struct{}
```
Comment (StephenButtolph) [inline utils/iterator/empty.go:13]: Would it make more sense to just have:
```suggestion
// Empty is an iterator with no elements.
type Empty[T any] struct{}
```
Doesn't seem like the `New` function buys us anything

File: utils/iterator/iteratormock/mock_iterator.go
Comment (StephenButtolph) [inline utils/iterator/iteratormock/mock_iterator.go:1]: Can we name this file `iterator`?

File: utils/iterator/iteratormock/mock_iterator.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+package iteratormock
+
+import (
+	reflect "reflect"
+
+	gomock "go.uber.org/mock/gomock"
+)
+
+// MockIterator is a mock of Iterator interface.
+type MockIterator[T any] struct {
```
Comment (StephenButtolph) [inline utils/iterator/iteratormock/mock_iterator.go:19]: Is it possible to name this `Iterator`?

File: utils/iterator/masked.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+var _ Iterator[any] = (*masked[any])(nil)
+
+type masked[T any] struct {
+	parentIterator Iterator[T]
+	filter         func(T) bool
+}
+
+// NewMasked returns a new iterator that skips the elements in [parentIterator]
+// that if the element is in [filter].
+func NewMasked[T any](parentIterator Iterator[T], filter func(T) bool) Iterator[T] {
```
Comment (StephenButtolph) [inline utils/iterator/masked.go:15]: I feel like `Filter` might be a better name. `iterator.Filter(myIterator, filter)`
```suggestion
// Filter returns an iterator that skips the elements in [it] that return true
// from [filter].
func Filter[T any](it Iterator[T], filter func(T) bool) Iterator[T] {
```

File: utils/iterator/masked.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package iterator
+
+var _ Iterator[any] = (*masked[any])(nil)
+
+type masked[T any] struct {
```
Comment (StephenButtolph) [inline utils/iterator/masked.go:8]:
```suggestion
type filtered[T any] struct {
```

File: utils/iterator/merged.go [hunk 5]
```diff
@@ -1,58 +1,63 @@
...
 	// exhausted.
-	heap heap.Queue[StakerIterator]
+	heap heap.Queue[Iterator[T]]
 }
 
-// Returns an iterator that returns all of the elements of [stakers] in order.
-func NewMergedIterator(stakers ...StakerIterator) StakerIterator {
+// NewMerged returns an iterator that returns all of the elements of [iterators]
+// in order.
+func NewMerged[T any](less btree.LessFunc[T], iterators ...Iterator[T]) Iterator[T] {
```
Comment (StephenButtolph) [inline utils/iterator/merged.go:23]:
```suggestion
// Merge returns an iterator that returns all of the elements of [iterators] in
// order.
func Merge[T any](less btree.LessFunc[T], iterators ...Iterator[T]) Iterator[T] {
```

File: utils/iterator/slice.go [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+var _ Iterator[any] = (*slice[any])(nil)
+
+type slice[T any] struct {
+	index    int
+	elements []T
+}
+
+// NewSlice returns an iterator that contains [elements] in order. Doesn't sort
+// by anything.
+func NewSlice[T any](elements ...T) Iterator[T] {
```
Comment (StephenButtolph) [inline utils/iterator/slice.go:15]:
```suggestion
// FromSlice returns an iterator that contains [elements] in order. Doesn't sort
// by anything.
func FromSlice[T any](elements ...T) Iterator[T] {
```

File: utils/iterator/tree.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+	current     T
+	next        chan T
+	releaseOnce sync.Once
+	release     chan struct{}
+	wg          sync.WaitGroup
+}
+
+// NewTree returns a new iterator of the stakers in [tree] in ascending order.
+// Note that it isn't safe to modify [tree] while iterating over it.
+func NewTree[T any](btree *btree.BTreeG[T]) Iterator[T] {
```
Comment (StephenButtolph) [inline utils/iterator/tree.go:24]:
```suggestion
// FromTree returns a new iterator of the stakers in [tree] in ascending order.
// Note that it isn't safe to modify [tree] while iterating over it.
func FromTree[T any](btree *btree.BTreeG[T]) Iterator[T] {
```

File: utils/iterator/empty.go [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+// See the file LICENSE for licensing terms.
+
+package iterator
+
+import "github.com/ava-labs/avalanchego/utils"
+
+// Empty is an iterator with no elements.
+type Empty[T any] struct{}
+
+func (Empty[T]) Next() bool {
```
Comment (StephenButtolph) [inline utils/iterator/empty.go:11]:
```suggestion
func (Empty[_]) Next() bool {
```

File: utils/iterator/empty.go [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+	return false
+}
+
+func (Empty[T]) Value() T {
+	return utils.Zero[T]()
+}
+
+func (Empty[T]) Release() {}
```
Comment (StephenButtolph) [inline utils/iterator/empty.go:19]:
```suggestion
func (Empty[_]) Release() {}
```

File: utils/iterator/filter.go [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+// Filter returns an iterator that skips the elements in [it] that return true
+// from [filter].
+func Filter[T any](it Iterator[T], filter func(T) bool) Iterator[T] {
+	return &filtered[T]{
+		it:     it,
+		filter: filter,
+	}
+}
+
+func (i *filtered[T]) Next() bool {
```
Comment (StephenButtolph) [inline utils/iterator/filter.go:22]:
```suggestion
func (i *filtered[_]) Next() bool {
```

File: utils/iterator/filter.go [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+		}
+	}
+	return false
+}
+
+func (i *filtered[T]) Value() T {
+	return i.it.Value()
+}
+
+func (i *filtered[T]) Release() {
```
Comment (StephenButtolph) [inline utils/iterator/filter.go:36]:
```suggestion
func (i *filtered[_]) Release() {
```

File: utils/iterator/merge.go [hunk 12]
```diff
@@ -1,58 +1,63 @@
...
-			stakers...,
+			iterators...,
 		),
 	}
 
 	return it
 }
 
-func (it *mergedIterator) Next() bool {
+func (it *merged[T]) Next() bool {
```
Comment (StephenButtolph) [inline utils/iterator/merge.go:52]:
```suggestion
func (it *merged[_]) Next() bool {
```

File: utils/iterator/merge.go [hunk 13]
```diff
@@ -71,18 +76,18 @@ func (it *mergedIterator) Next() bool {
...
 }
 
-func (it *mergedIterator) Value() *Staker {
+func (it *merged[T]) Value() T {
 	peek, _ := it.heap.Peek()
 	return peek.Value()
 }
 
-func (it *mergedIterator) Release() {
+func (it *merged[T]) Release() {
```
Comment (StephenButtolph) [inline utils/iterator/merge.go:84]:
```suggestion
func (it *merged[_]) Release() {
```

File: utils/iterator/merge.go [hunk 14]
```diff
@@ -71,18 +76,18 @@ func (it *mergedIterator) Next() bool {
...
 	for it.heap.Len() > 0 {
 		removed, _ := it.heap.Pop()
 		removed.Release()
 	}
 }
 
-func (it *mergedIterator) Len() int {
+func (it *merged[T]) Len() int {
 	return it.heap.Len()
 }
```
Comment (StephenButtolph) [inline utils/iterator/merge.go:93]: When is this used?
```suggestion
```

File: utils/iterator/slice.go [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+// FromSlice returns an iterator that contains [elements] in order. Doesn't sort
+// by anything.
+func FromSlice[T any](elements ...T) Iterator[T] {
+	return &slice[T]{
+		index:    -1,
+		elements: elements,
+	}
+}
+
+func (i *slice[T]) Next() bool {
```
Comment (StephenButtolph) [inline utils/iterator/slice.go:22]:
```suggestion
func (i *slice[_]) Next() bool {
```

File: utils/iterator/slice.go [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+	i.index++
+	return i.index < len(i.elements)
+}
+
+func (i *slice[T]) Value() T {
+	return i.elements[i.index]
+}
+
+func (*slice[T]) Release() {}
```
Comment (StephenButtolph) [inline utils/iterator/slice.go:31]:
```suggestion
func (*slice[_]) Release() {}
```

File: utils/iterator/tree.go [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+			case <-it.release:
+				return false
+			}
+		})
+		close(it.next)
+	}()
+	return it
+}
+
+func (i *tree[T]) Next() bool {
```
Comment (StephenButtolph) [inline utils/iterator/tree.go:48]:
```suggestion
func (i *tree[_]) Next() bool {
```

File: utils/iterator/tree.go [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+	next, ok := <-i.next
+	i.current = next
+	return ok
+}
+
+func (i *tree[T]) Value() T {
+	return i.current
+}
+
+func (i *tree[T]) Release() {
```
Comment (StephenButtolph) [inline utils/iterator/tree.go:58]:
```suggestion
func (i *tree[_]) Release() {
```

**Review by StephenButtolph — COMMENTED** [review]
nits then lgtm

File: utils/iterator/merge.go (hunk 14 above)
Comment (dhrubabasu) [reply utils/iterator/merge.go:93]: it isn't, deleted

**Review by StephenButtolph — APPROVED** [review]

**Pull request merged**